components/remote_executor/geoapp_client.py

```diff
@@ -0,0 +1,128 @@
+RETRY_LIMIT_SECONDS = 172800.0  # 2 days in seconds
+REQUEST_TIMEOUT = 10
+
+
+class GeoappClient:
+    def __init__(self, geoap_creds):
+
+        log = logging.getLogger('urllib3')
+        log.setLevel(logging.DEBUG)
+        log.addHandler(logging.StreamHandler())
+
+        retry_strategy = Retry(total=RETRY_INTERVAL,
+                               backoff_factor=60, status_forcelist=[502, 503, 504])
+        timeout = urllib3.Timeout(
+            total=RETRY_LIMIT_SECONDS, connect=10.0, read=10.0)
+        http = urllib3.PoolManager(timeout=timeout, retries=retry_strategy)
+
+        with open(geoap_creds, 'r') as f:
+            geoap_creds_data = json.load(f)
+
+        self.api_endpoint = geoap_creds_data["API_ENDPOINT"]
+        self.http = http
+        self.log = log
+        self.params = {"apikey": geoap_creds_data["API_KEY"]}
+
+    def get_component_id(self, name):
+        url = self.api_endpoint + "api/notebook"
+        response = self.http.request('GET', url, fields=self.params)
+        if response.status != 200:
+            raise Exception(
+                f"Get Component ID error  - status code {response.status_code}")
+
+        components_list = json.loads(response.data.decode())
+        for component in components_list:
+            if component.get("name", "") == name:
+                component_id = component.get("id")
+                self.log.info(f"component_id: {component_id}")
+                return component_id
+        return None
+
+    def get_user_id(self):
+        url = self.api_endpoint + "api/users/current"
+        response = self.http.request('GET', url, fields=self.params)
+        if response.status != 200:
+            raise Exception(
+                f"Get User ID error  - status code {response.status_code}")
+        user_id = json.loads(response.data.decode()).get("pk")
+        self.log.info(f"user_id: {user_id}")
+        return user_id
+
+    def create_request(self, data):
+        url = self.api_endpoint + "api/request"
+        response = self.http.request(
+            'POST', url + f"?apikey={self.params['apikey']}", fields=data)
+        if response.status not in (200, 201):  # TODO Left only 201
+            raise Exception(
+                f"Create Request error - status code {response.status_code}")
+        request = json.loads(response.data.decode())
+        self.log.info(f"request: {request}")
+        return request
+
+    def wait_for_request_success_finish(self, request_id):
+        url = self.api_endpoint + f"api/request/{request_id}"
+        start_time = time.time()
+        self.log.info(
+            f"wait_for_request_success started: {time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(start_time))}")
+        while time.time() < start_time + RETRY_LIMIT_SECONDS:
+            response = self.http.request('GET', url, fields=self.params)
+            curr_request = json.loads(response.data.decode())
+            if not curr_request.get("calculated"):
```

Comment:
Look for finished_at. If it is set then look for calculated

Reply:
done